processing/src/main/java/org/apache/druid/query/filter/EqualityFilter.java

```diff
@@ -0,0 +1,578 @@
+    }
+
+    final StringValueSetIndex valueSetIndex = indexSupplier.as(StringValueSetIndex.class);
+    if (valueSetIndex == null) {
+      // column exists, but has no index
+      return null;
+    }
+    return valueSetIndex.forValue(String.valueOf(matchValue));
+  }
+
+  @Override
+  public ValueMatcher makeMatcher(ColumnSelectorFactory factory)
+  {
+    return ColumnProcessors.makeProcessor(
+        column,
+        new TypedConstantValueMatcherFactory(matchValue, matchValueType),
+        factory
+    );
+  }
+
+  @Override
+  public VectorValueMatcher makeVectorMatcher(VectorColumnSelectorFactory factory)
+  {
+    final ColumnCapabilities capabilities = factory.getColumnCapabilities(column);
+
+    if (matchValueType.isPrimitive() && (capabilities == null || capabilities.isPrimitive())) {
+      return ColumnProcessors.makeVectorProcessor(
+          column,
+          VectorValueMatcherColumnProcessorFactory.instance(),
+          factory
+      ).makeMatcher(matchValue, matchValueType);
+    }
+    return ColumnProcessors.makeVectorProcessor(
+        column,
+        VectorValueMatcherColumnProcessorFactory.instance(),
+        factory
+    ).makeMatcher(new EqualityPredicateFactory(matchValue, matchValueType));
+  }
+
+  @Override
+  public boolean supportsSelectivityEstimation(ColumnSelector columnSelector, ColumnIndexSelector indexSelector)
+  {
+    return Filters.supportsSelectivityEstimation(this, column, columnSelector, indexSelector);
+  }
+
+  @Override
+  public boolean canVectorizeMatcher(ColumnInspector inspector)
+  {
+    return true;
+  }
+
+  @Override
+  public Set<String> getRequiredColumns()
+  {
+    return ImmutableSet.of(column);
+  }
+
+  @Override
+  public boolean supportsRequiredColumnRewrite()
+  {
+    return true;
+  }
+
+  @Override
+  public Filter rewriteRequiredColumns(Map<String, String> columnRewrites)
+  {
+    String rewriteDimensionTo = columnRewrites.get(column);
+
+    if (rewriteDimensionTo == null) {
+      throw new IAE(
+          "Received a non-applicable rewrite: %s, filter's dimension: %s",
+          columnRewrites,
+          columnRewrites
+      );
+    }
+
+    return new EqualityFilter(
+        rewriteDimensionTo,
+        matchValueType,
+        matchValue,
+        extractionFn,
+        filterTuning
+    );
+  }
+
+  private static class EqualityPredicateFactory implements DruidPredicateFactory
+  {
+    private final ExprEval<?> matchValue;
+    private final ColumnType matchValueType;
+
+    private final Object initLock = new Object();
+
+    private volatile DruidLongPredicate longPredicate;
+    private volatile DruidFloatPredicate floatPredicate;
+    private volatile DruidDoublePredicate doublePredicate;
+
+    public EqualityPredicateFactory(Object matchValue, ColumnType matchValueType)
+    {
+      this.matchValue = ExprEval.ofType(ExpressionType.fromColumnType(matchValueType), matchValue);
+      this.matchValueType = matchValueType;
```

Comment:
I'm a little concerned that this code is trying to force materialize something that it doesn't need to force materialize.  I'm a bit confused as to how this would work with something like a sketch or other complex object where we are passing in an object that the column would know how to deal with, but the expression evaluation system doesn't.  I'm scared that we will have to teach expressions about how to deal with all of the types when we've already taught the column how to deal with them and all we really want is the Object to be passed through un-changed such that the column can do the right thing.

So, all of that said, to say, if I'm a column implementation of a relatively complex object (let's say HllSketch) and I want to support this Equality operator (I want someone to be able to pass in the base64 encoded string of the HllSketch and I want to find the rows that match it), how can I do that?  Or do I need to create my own "HllSketchFilter" and use that instead?

Reply:
>I'm a little concerned that this code is trying to force materialize something that it doesn't need to force materialize. I'm a bit confused as to how this would work with something like a sketch or other complex object where we are passing in an object that the column would know how to deal with, but the expression evaluation system doesn't. I'm scared that we will have to teach expressions about how to deal with all of the types when we've already taught the column how to deal with them and all we really want is the Object to be passed through un-changed such that the column can do the right thing.

I think your concern might be with mixing up `ExprEval.ofType` with `ExprEval.bestEffortOf`, the latter of which blindly tries to find the best type and is used by the 'auto' indexers. `ExprEval.ofType` what is used for expression bindings backed by columns at query time, and is a fairly light wrapper. When used with complex types, it has some special handling when the value is `String` (which is assumed to be base64 and decoded into `byte[]`) or `byte[]` which will be fed to the complex serde to turn into objects, but otherwise is basically a wrapper that holds type information but otherwise does nothing to the thing it is holding (this is how expressions that operate on complex columns are able to do their thing as well). This is also a large part of why `matchValueType` is specified here, for future expandability.

> So, all of that said, to say, if I'm a column implementation of a relatively complex object (let's say HllSketch) and I want to support this Equality operator (I want someone to be able to pass in the base64 encoded string of the HllSketch and I want to find the rows that match it), how can I do that? Or do I need to create my own "HllSketchFilter" and use that instead?

Right now complex types do not define an equality method, only a sorting comparator, but I imagine a future where we add this to the type serde (or something), which would allow equality comparison for arbitrary complex types. Since complex types right now end up using the object processor/predicate, equality is only currently supported for things which can work with `Objects.equals`, however, if we add this mechanism I think it would be pretty easy to allow all complex types to opt-in to supporting this, and replace the current special handling i have to support nested column equality https://github.com/apache/druid/pull/14542/files#diff-b5d0b80ff07d3a5ceaa2a651d619216fcd09d77805f9412d09935a912db774f4R412

Reply:
> Since complex types right now end up using the object processor/predicate, equality is only currently supported for things which can work with Objects.equals, however, if we add this mechanism I think it would be pretty easy to allow all complex types to opt-in to supporting this

Is this really true?  The complex columns themselves get to decide how equality is defined because they return their own implementations of `StringValueSetIndex` (and eventually `ObjectValueSetIndex`, when we get there).  The `Object.equals` thing is just a hold out because we haven't moved the predicates to be bundled with the indexes returned from the various semantic interfaces that are passed in the `.as()` method.

Reply:
talked about this offline, but for the sake of everyone else, it is true for the value matchers, at least given how things work today. The addition of `TypedValueIndex` in this PR does push down the value index selection and the native match value type down to the column index supplier so that equality check can be done in whatever manner makes sense for the column, but quite a heavy refactor is required to make this true of the value matchers.